Package manager selection

tristonarmstrong opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/futurizeworld/mantis-cli/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #32 


## What is the new behavior?
Adds potentially 2 new steps: minimum 1 extra step.
- step 1: after the database connection question, the user is prompted with `Would you like to auto detect the package manager used to install the dependencies?` (default = yes)
   - This is done to allow the user the option to skip defining a package manager
-  step 2: (if step 1 is no) then the user is prompted with a list of package managers predefined by mantisCLI
   - performs install as usual

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
There doesn't seem to be any tests added to this project yet, as well as the `npm run clean` command has been removed in a previous commit. The `test` command fails as there's no test directory or jest config file

tristonarmstrong commented:
Consider this PR the opening of the discussion. I hope this satisfies the direction intended, if not, feel free to let me know and ill make adjustments.

tristonarmstrong commented:
Commit signing was missed, Ill squash again and sign the commit

tristonarmstrong committed:
feat(init.command.ts): add support for user selected pm

Auto detect functionality option Isn't necessarily in
the issue ticket, but i think it makes sense to give users
the option to not have to think too much about it.

Open to suggestion

Resolves #32

RyanClementsHax commented:
Thanks for the contribution! I really like what you've done.

I see the flow now looks something like this:
```txt
Would you like to auto detect the package manager used to install the dependencies?
[ ] Yes
[x] No

Select the package manager you'd like to use:
[ ] npm
[ ] yarn
[ ] pnpm
[ ] bun
```

What do you think of this?
```txt
What package manager would you like to use?
[ ] npm (template default)
[ ] yarn
[ ] pnpm
[ ] bun
```

This builds on top of what you've done but also gives the user a hint of what the "default package manager" is for the template selected. This does mean the detected package manager should be sorted as the first item in the list so it can be selected by default.

Also what do you think of deleting the lockfile in the generated repo if a different package manager is selected?

tristonarmstrong commented:
I like that, consolidates the two options and removes the need for the whole "auto detect" stuff.

Tho, about the lock file... Looks like we've got some things to consider. 
1) bun doesnt seem to generate a `bun.lockb` when using the `--frozen-lockfile` flag and the `package.lock` file exists. Bun attempts to migrate the `package-lock` from npm to `bun.lockb`. I guess this TECHNICALLY isnt an issue since bun will just migrate the existing lock file every time? but nice to be aware of. Bun will also error out if using `--frozen-lockfile` and no lockfile exists which means we would obviously have to delete the lock file post migration and not pre lol -> [bun docs](https://bun.sh/docs/runtime/bunfig#install-frozenlockfile)
2) pnpm completely errors out when using the `--frozen-lockfile` flag as theres not already a `pnpm-lock.yaml` file present & valid -> [pnpm docs](https://pnpm.io/cli/install#--frozen-lockfile)
3) yarn SEEMS to ignore the existing lock file and installs from scratch? Also not creating a yarn lock when using the `--frozen-lockfile` flag, even if initial `package-lock` isnt present. Using `yarn import` attempts to migrate the `package-lock` over to `yarn.lock`, tho fails with the template this project uses [ERROR: *source files corrupted*] -> [yarn docs v1](https://classic.yarnpkg.com/en/docs/cli/import)

These issues ofcourse go away if you dont use the `--frozen-lockfile` flag

---
embarrassingly didn't catch any of this during my initial stab at this PR. Just made the wild assumption that it'd all just magically work

Commit:
Merge branch 'main' into package-manager-selection

RyanClementsHax commented:
@tristonarmstrong Good finds!

You're right that the package manager detection/install commands are half baked and only work with `npm`. For context, I added this plumbing in anticipation for a PR like this to drop one day.

We have a few options that I see.

1. Change the template to not have a lock file and have the install command generate one based on package manager selection
    - Pros:
        - Makes it easy to implement package management selection
    - Cons:
        - Makes installs non-deterministic
2. Change the template to have a lock file for each package manager
    - Pros:
        - Allows installs to be deterministic while maintaining the package manager selection feature
    - Cons:
        - Makes creating templates harder
3. Remove package manager selection all together
    - Pros:
        - Makes things simpler
        - Makes installs deterministic and easier to debug
    - Cons:
        - Need to remove the package manager selection feature
        
For comparison, neither nuxt, analog, nor next project generators use lock files. As such I'm inclined toward option 1. What do you think @tristonarmstrong @adyngom ?

tristonarmstrong commented:
Im not apposed to scrapping this PR in favor of a simpler, more deterministic approach. Tho, the Bun lover in me would really hate to sit through an npm install lol Thus my vote would go toward option 1

RyanClementsHax commented:
Let's go with option 1 then. That way we can satisfy the bun lover in you lol.

Is this something you still want to contribute?

tristonarmstrong commented:
absolutely. I'll allocate some time for it today, to get that done.

just so we're on the same page:
- [x] remove lock file in template
- [x] remove `--frozen-lock` flag on command map
- [x] remove "auto detect" prompt/logic

tristonarmstrong committed:
feat(init.command.ts): change package manager selection/install logic

- remove the `--frozen-lockfile` flag
- remove the auto detect package manager function and usage

tristonarmstrong committed:
feat: remove lock file from template

We are no longer generating templates with an existing lock file

tristonarmstrong commented:
Did not include the `(template default)` label on the `npm` selection since the templates will no longer have a lock file.

tristonarmstrong committed:
feat(init.command.ts): replace ci command with install

npm ci can only be run if theres an existing lock file

tristonarmstrong commented:
All seems to be fully functional EXCEPT `yarn` & `pnpm`. I get the following error when starting application:

yarn
``` bash
failed to load config from /Users/triston/Documents/mantis-cli/bin/mantis-todo/apps/web-client/vite.config.ts
 NX   Cannot find package '@ngtools/webpack' imported from /Users/triston/Documents/mantis-cli/bin/mantis-todo/node_modules/@analogjs/vite-plugin-angular/src/lib/host.js
```

pnpm (doesnt occur if i run `<otherPM> install` right after)
```bash
Error: apps/mobile-client/src/app/home/home.page.ts:2:63 - error TS2307: Cannot find module '@ionic/core' or its corresponding type declarations.

2 import { IonRefresherCustomEvent, RefresherEventDetail } from '@ionic/core';
```
^^ setting this as a peer dependency seems to do the trick

RyanClementsHax commented:
Ahhhh, that makes sense. The package managers have odd behaviors between each other when it comes to indirect dependencies.

Including `@ngtools/webpack` as a dev dependency should be fine to solve the first issue.
Including `@ionic/core` as a direct dependency should be fine to solve the second issue.

RyanClementsHax commented on src/commands/init.command.ts:
```diff
@@ -157,20 +158,37 @@ const renameGitignoreFiles = async (workspacePath: string) => {
 const pmToInstallCommandMap: Record<PackageManagerName, [string, string[]]> = {
-  npm: ['npm', ['ci']],
-  yarn: ['yarn', ['install', '--frozen-lockfile']],
-  bun: ['bun', ['install', '--frozen-lockfile']],
-  pnpm: ['pnpm', ['install', '--frozen-lockfile']],
+  npm: ['npm', ['install']],
+  yarn: ['yarn', ['install']],
+  bun: ['bun', ['install']],
+  pnpm: ['pnpm', ['install']],
+};
```
Good news! You can delete all of this and use [installDependencies(...)](https://github.com/unjs/nypm?tab=readme-ov-file#installdependenciesoptions) directly! Hooray for deleting code!

tristonarmstrong committed:
feat(templates/*): add packages to pacakge.json in template

this is done to satisfy some of the nuances that some of the other
package managers seem to have

tristonarmstrong committed:
feat(init.command.ts): Use installDependencies in favor of execa

- remove the `execa` function call in favor of nypm's `installDependencies` function
- remove the
`pmToInstallCommandMap` in favor of just a list of package manager names

issue #32

tristonarmstrong committed:
feat(init.command.ts): Remove unnecessary guard clause w/ pm selection

issue #32

tristonarmstrong commented on src/commands/init.command.ts:
```diff
@@ -155,27 +156,29 @@ const renameGitignoreFiles = async (workspacePath: string) => {
+    type: 'select',
+    name: 'name' satisfies keyof PMTypePromptResult,
+    message: "Select the package manager you'd like to use:",
+    choices: ['npm', 'bun', 'pnpm', 'yarn'] satisfies PackageManagerName[],
```
so `nypm` exports a `packageManagers` list of `PackageManager[]` type. The only issue is, it contains two major versions on `yarn (1.x,3.x)`. Hard defining the list of packages removes the need for filtering out that particular package from displaying in the list of packages. We COULD use the hint property on the choices list items to annotate the version number next to both yarns which then would require a slight change in the return types and whats passed to `installDependencies`, but thats no biggie

At the moment doing it this way keeps it consistent with your original intentions, i believe

tristonarmstrong commented on src/commands/init.command.ts:
```diff
@@ -155,27 +156,29 @@ const renameGitignoreFiles = async (workspacePath: string) => {
+  const pm = await promptForPackageManagerSelect();
   const spinner = ora('Installing dependencies').start();
-  const pm = await detectPackageManager(workspacePath);
-  if (!pm) {
-    throw new Error(
-      'No package manager found in the workspace. Unable to install dependencies.',
-    );
-  }
```
I removed this pm check because it seems to me that it was used when we were auto detecting the package manager. Now that we offer a defined list of package managers it seems redundant to me.

RyanClementsHax reviewed (approved):
Great work. I miss working with you 💔
